scaling.tex

```diff
@@ -1 +1,216 @@
+code, in this case kernels such as the various matrix-vector
+routines. The Halo Exchanges (labelled ``HE'') appear in the {\em ETC}
+category as YAXT is compiled as a library, in this case without the
+CrayPAT profiling, so it is unable to see that this ultimately calls
+MPI. The Global Sums (labelled ``GS'') appear under {\em MPI} as the
+global summation calls are in the LFRic infrastructure code.
+
+The strong scaling of the user code is shown in
+figure~\ref{fig:U_scale}\footnote{There is no hybrid data for 24 nodes
+  as the model didn't run to completion so no profile was generated.}.
+Both the MPI only and hybrid code show excellent scaling - the
+communication costs are represented in other categories. For MPI
+only, comparing the $24$ node to the $216$ node jobs, the User code
+runs nearly $9\times$ faster for $9$ times the resource. The hybrid
+version fares even better, showing super-linear scaling. Comparing the
+$54$ node job to the $384$ node job, the user code runs more than
+$12\times$ faster for $7.11\times$ the resource. This may well be due
+to caching effects, once the local volume becomes small enough to fit
+into cache.
+
+
+\begin{figure}[ht!]
+\centering\includegraphics[width=1.0\linewidth]{figs/U-scale.eps}
+\caption{\label{fig:U_scale}Wall-clock time strong scaling of the 
+  User code from Baroclinic wave test.}
+\end{figure} 
+
+Shown in figure~\ref{fig:comms_scale} and Table~\ref{tab:scale-data}
+is the strong scaling of the communication components as measured by
+the CrayPAT tool. The halo exchanges for hybrid version of the code
+takes significantly longer than the MPI only version. The time
+difference accounts for the hybrid version being slower than MPI only.
+The count of the number of halo exchanges (performed internally by
+LFRic) is almost the same for the MPI and hybrid versions. The
+difference is a few hundred out of a total of $1.3$ million. The are
+not exactly the same because the global sum is not reproducible across
+different partitions, so the evolution is slightly different. However,
+after $100$ time-steps and many thousands of solver iterations, the
+conservative fields are still the same to many orders of magnitude.
+
+\begin{figure}[ht!]
+\centering\includegraphics[width=1.0\linewidth]{figs/comms-scale.eps}
+\caption{\label{fig:comms_scale}Wall-clock time strong scaling of the 
+  communications costs of the Baroclinic wave test. HE denotes Halo
+  Exchange and GS the Global Sum.}
+\end{figure} 
+
+The size of the halos is shown in Table~\ref{tab:scale-data}. The 
+halos for the hybrid version are bigger than the MPI only version, as 
+there are less MPI ranks, but there are less messages to send. As the 
+hybrid version is running with six threads, it has six times less the 
+number of messages. The hybrid version has fewer, but bigger MPI 
+messages and moreover, will have less total data to send. It is not 
+obvious why it should be so much slower. 
```

Comment:
Just a thought about the cause of the slowdown- have we reached a message size threshold? What do we set $MPICH_GNI_MAX_EAGER_MSG_SIZE to?

Reply:
Yes, could be this. Not set to anything, so just the default. It was one of the things on my list to investigate. Do you know what the thresholds are?

Reply:
I don't but I'll send you an email with the settings that we use in UM production runs.